godot_code/main.gd

```diff
@@ -1,34 +1,61 @@
 extends Control
 
-@onready var menuPlace: Control = $MainWindow/Columns/MenuCol
+@onready var mainMenuWindow: VBoxContainer = $MainWindow
+@onready var controller: Node = $GameNiteControlpads
+var gameNode: Node = null
+
+var gameState: String = "mainMenu"
+var players: Dictionary = {}
 
-# Called when the node enters the scene tree for the first time.
 func _ready():
-	menuPlace.get_child(0).connect("startButtonPressed", openGameJoin)
-
-# Called every frame. 'delta' is the elapsed time since the previous frame.
-func _process(delta):
-	pass
-
-func openGameJoin():
-	var lobbyScene = preload("res://player_lobby.tscn")
-	for child in menuPlace.get_children():
-		child.queue_free()
-	var lobbyNode: VBoxContainer = lobbyScene.instantiate()
-	lobbyNode.connect("backToMain", openMainMenu)
-	menuPlace.add_child(lobbyNode)
-
-func openMainMenu():
-	var menuScene = preload("res://main_menu.tscn")
-	for child in menuPlace.get_children():
-		child.queue_free()
-	var menuNode: VBoxContainer = menuScene.instantiate()
-	menuNode.connect("startButtonPressed", openGameJoin)
-	menuPlace.add_child(menuNode)
-
-func _on_game_nite_controlpads_message_received(client, message):
-	print("received <%s> from <%s>" % [message, client])
-
-
-func _on_exit_button_pressed():
-	get_tree().quit()
+	mainMenuWindow.connect("stateChanged", stateChanged)
+
+func stateChanged(newState: String):
+	match newState:
+		"startGame":
+			assert(gameState != "startGame")
+			assert(gameNode == null)
+			var gameScene = preload("res://game/game.tscn")
+			var newGameNode: Control = gameScene.instantiate()
+			gameNode = newGameNode
+			add_child(gameNode)
+	gameState = newState
+
+func playerJoins(clientId: String, playerName: String):
+	if gameState != "joinGame":
+		print_debug("can't join now")
+		sendMessage(clientId, "JoinFailed", {"status": "notAvailable"})
+		return
+	if clientId in players:
```

Comment:
I think we should add a check for duplicated names, since if two players have the same name it will be hard to pick from a list